### ISSUE TITLE
AssemblyLoadContext - WPF - Collectable Assemblies are not garbage collected

### ISSUE THREAD

**Issue by X1C3** [dotnet/coreclr#26055]
I have a problem with a plugin that has a wpf usercontrol. If an instance of the control is created then the plugin assembly will not be garbage collected. Whats wrong here?

```
Project1:
public interface IPlugin
    {
        object GuiControl { get; }
    }
```

```
Project2:
public class Plugin : IPlugin
    {
        public object GuiControl { get; }

        public Plugin()
        {
            GuiControl = new UserControl1(); // Comment this line out => Assemblies are now garbage collected
        }

        ~Plugin()
        {
            Console.WriteLine("Plugin: Calling destructor");
        }
    }
```

```
Project3:
public class CollectableAssemblyLoadContext : AssemblyLoadContext
    {
        public CollectableAssemblyLoadContext(string name) : base(name, true) { }
              
        protected override Assembly Load(AssemblyName assemblyName) => null;

        public Assembly Load(byte[] rawAssembly)
        {
            using (var memoryStream = new MemoryStream(rawAssembly))
            {
                var assembly = LoadFromStream(memoryStream);
                return assembly;
            }
        }
    }

class Program
    {
        [STAThread]
        static void Main(string[] args)
        {
            for (int i = 0; i < 500; i++)
                TestRun();

            int counter = 0;
            //while(counter == 100)
            //{
            counter++;

            GC.Collect();
            GC.WaitForPendingFinalizers();

            //Thread.Sleep(2500);
            //}

            Console.WriteLine("Assemblies: {0}", AppDomain.CurrentDomain.GetAssemblies().Where(p => p.IsCollectible).Count()); // 500 Assemblies - Why?
            Console.WriteLine("ALC's: {0}", AssemblyLoadContext.All.Count()); // 1 ALC "Default"

            Console.ReadKey();
        }       
        
        public static void TestRun()
        {
            var context = new CollectableAssemblyLoadContext(Guid.NewGuid().ToString());
            var assembly = context.LoadFromAssemblyPath(Path.Combine(Directory.GetCurrentDirectory(), "GuiPlugin.dll"));

            var instance = (IPlugin) Activator.CreateInstance(assembly.GetType("GuiPlugin.Plugin"));            

            context.Unload();
        }
    }
```

**Comment by mattwarren** [comment]
Out of interest, if you add `[MethodImpl(MethodImplOptions.NoInlining)]` to your `TestRun()` method, does that make any difference?

I've seen that attribute used in a several examples, as it helps prevent assemblies being inadvertently 'rooted', e.g.

- [The official docs](https://docs.microsoft.com/en-us/dotnet/standard/assembly/unloadability-howto#example-source-with-unloadability-issues)
- https://github.com/dotnet/coreclr/issues/20694

**Comment by AndyAyersMS** [comment]
It's possible that the jit might extend GC lifetimes of inlinee locals beyond the extent of the inlinee. #9479 tried to put a stop to this, so if you find examples where this happens I'd be happy to take a look.

For allocations within the same method where you force GC, even if you null out your locals after using them, the jit may have stashed references in jit temporaries that don't get nulled.

Keeping your expected to be dead allocations and references in a noinline method is the most reliable fix.

**Comment by AaronRobinsonMSFT** [comment]
cc @janvorli

**Comment by X1C3** [comment]
> Out of interest, if you add `[MethodImpl(MethodImplOptions.NoInlining)]` to your `TestRun()` method, does that make any difference?
> 
> I've seen that attribute used in a several examples, as it helps prevent assemblies being inadvertently 'rooted', e.g.
> 
> * [The official docs](https://docs.microsoft.com/en-us/dotnet/standard/assembly/unloadability-howto#example-source-with-unloadability-issues)
> * #20694

Adding `[MethodImpl(MethodImplOptions.NoInlining)]` to the method `TestRun()` did not help.

**Comment by X1C3** [comment]
It's strange ... if i am change
`GuiControl = new UserControl1(); // Comment this line out => Assemblies are now garbage collected`
to
`GuiControl = new UserControl(); // Comment this line out => Assemblies are now garbage collected`

Assemblies are now garbage collected.

**Comment by janvorli** [comment]
@X1C3 how do UserControl and UserControl1 differ?

**Comment by X1C3** [comment]
@janvorli

UserControl1 inherit from UserControl. UserControl1 has no special code implementation. It is empty.

**Comment by janvorli** [comment]
@X1C3 I have just noticed you only call GC.Collect(); GC.WaitForPendingFinalizers(); once. That doesn't guarantee that the unload will complete after that. If you want to make sure that everything is unloaded at some specific point, you need to do the following (as described in the https://docs.microsoft.com/en-us/dotnet/standard/assembly/unloadability-howto#example-source-with-unloadability-issues).
So in your test, you'd need to:
* Add [MethodImpl(MethodImplOptions.NoInlining)] to the TestRun
* Change the TestRun to return WeakReference and return new WeakReference(context) from it
* In your Main, have a loop that calls GC.Collect(); GC.WaitForPendingFinalizers(); while the weak reference is alive. Since you call the RunTest 500 times in a loop, you could either do it after each call to TestRun or have an array of 500 WeakReferences that you set while calling the RunTest and then have a loop that calls the GC.Collect(); GC.WaitForPendingFinalizers(); until all of the WeakReferences become invalid.

**Comment by X1C3** [comment]
@janvorli 

I think that the problem is not the implementation, but the wpf control (UserControl1). Without this UserControl1 the assembly load context and the loaded assembly is garbage collected immediately.

```
[STAThread]
        static void Main(string[] args)
        {
            //for (int i = 0; i < 500; i++) 

            WeakReference weakReference;
            TestRun(out weakReference);

            for (int i = 0; weakReference.IsAlive && (i < 10000); i++)
            {
                GC.Collect();
                GC.WaitForPendingFinalizers();
                Console.WriteLine("{0}: " + weakReference.IsAlive.ToString(), i);
            }                  
        }        

        [MethodImpl(MethodImplOptions.NoInlining)]
        public static void TestRun(out WeakReference weakReference)
        {
            var context = new CollectableAssemblyLoadContext(Guid.NewGuid().ToString());
            weakReference = new WeakReference(context);

            var assembly = context.LoadFromAssemblyPath(Path.Combine(Directory.GetCurrentDirectory(), "GuiPlugin.dll"));

            var type = assembly.GetType("GuiPlugin.Plugin");
            Activator.CreateInstance(type);
            type = null;

            context.Unload();
        }
```

**Comment by janvorli** [comment]
@X1C3 so do I understand it correctly that the code above fails to unload  with UserControl1 even with many iterations in the `for` loop?

**Comment by X1C3** [comment]
@janvorli 
Correct, the weakReference status is always "alive".

**Comment by janvorli** [comment]
@X1C3 I've tried to replicate this locally and I get the following output:
```
Plugin: Calling destructor
0: True
1: True
2: False
```
So it seems to work. I wonder - is your main app that you are testing it with a WPF app or a console app? I am testing it with a console one, so that may cause the difference.

**Comment by X1C3** [comment]
@janvorli  
Here a repro: https://github.com/X1C3/WpfPlugin

I am using a console app. I would like to know, what is different in your implementation.

**Comment by janvorli** [comment]
Ok, my local repro has the UserControl in a separate assembly. That makes the difference. I've just split your Plugin project into two - one with the Plugin.cs and the other with the UserControl1 files. 
I have not tried to look into why it makes a difference yet.

**Comment by X1C3** [comment]
@janvorli 
Thanks for the hint. Now it works if the UserControl is moved to a seperate assembly file. But that should not be the final solution.

**Comment by janvorli** [comment]
I agree that it is strange. I'll debug it to get full understanding of the different behavior.

**Comment by janvorli** [comment]
@X1C3 I have found why it works when the control is in a separate assembly. I have missed before that your custom CollectableAssemblyLoadContext doesn't resolve other assemblies than the plugin one into this context (the Load method returns null). So when the control is in a separate assembly, it is loaded into the default context and thus it doesn't prevent unloading of the context. But that's not what you wanted do end up with.

The real culprit is in the WPF itself. The MS.Internal.Resources.ResourceManagerWrapper holds on the assembly containing the control. And it is rooted (transitively) by two different roots. I've just digged that from the !gcroots command in WinDbg at the point where the unload fails.
My knowledge of WPF and its source code is zero, so I am not sure if there is a way to somehow explicitly remove the ResourceManagerWrapper for our specific assembly using existing APIs or not. But looking at the source code of MS.Internal.AppModel.ResourceContainer that manages those, it seems assemblies can be only added, but never removed.

So it seems WPF will need to be made aware of unloadability to make your scenario work.

**Comment by X1C3** [comment]
@janvorli
Thanks for analysing. Should this restriction fix by the wpf team? Is it kind of a bug or more an extension in wpf - what is the next step?

**Comment by janvorli** [comment]
Someone will need to fix this in the WPF repo (https://github.com/dotnet/wpf). Since it is open source, it doesn't have to be the wpf team.
I would not call it a bug, rather a missing support for unloading. The unloadability support was added quite recently and so many libraries will need to catch up. I'll create a github issue in the wpf repo for making WPF unloadability friendly.

**Comment by janvorli** [comment]
Created issue here: https://github.com/dotnet/wpf/issues/1764
